### PR TITLE
Deque store debug print

### PR DESCRIPTION
**Pull request by FSoaresDev**
This is a storage wrapper based on AppendStore called DequeStore.
It guarantees constant-cost appending to and popping from a list of items in storage on both directions (front and back).
Also adds a "remove" feature that receives an index and removes it from the list of items

### CODE REVIEW

File: packages/storage/src/deque_store.rs [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        assert_eq!(deque_store.remove(3), Ok(4));
+        assert_eq!(deque_store.get_at(4), Ok(6));
+        assert_eq!(deque_store.get_at(3), Ok(5));
+        assert_eq!(deque_store.get_at(2), Ok(3));
+        assert_eq!(deque_store.get_at(1), Ok(2));
+        assert_eq!(deque_store.get_at(0), Ok(1));
+
+        assert_eq!(deque_store.remove(1), Ok(2));
+        assert_eq!(deque_store.get_at(3), Ok(6));
+        assert_eq!(deque_store.get_at(2), Ok(5));
```
Comment (reuvenpo) [inline packages/storage/src/deque_store.rs:540]: Nice tests!
After doing the changes that i mentioned to `remove`, please make this deque 8 long, then start by trying to pop at `8` and `9` to see that it fails, then popping `0`, then `len -1`, then `1`, then `len - 2`, and then continue however you like until it's empty. That way I'll be sure that the function is behaving well.

**Review by reuvenpo — CHANGES_REQUESTED** [review]
Thank you for this PR!
This code looks good, but i have a few comments you can see below.
Additionally, i have some general notes:
- Make sure you run rustfmt on your code. you can either do it manually by running `cargo fmt`, or configure your IDE to do it automatically for you (both jetbrains IDEs and VSCode have a config option for it, when the rust plugin is installed)
- Avoid using `.unwrap()` wherever you can. The code should not panic as that causes the contract to halt and the transaction to fail with no indication to the user about what went wrong

File: packages/storage/src/deque_store.rs [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+{
+    /// Try to use the provided storage as an DequeStore. If it doesn't seem to be one, then
+    /// initialize it as one. This method allows choosing the serialization format you want to use.
+    ///
+    /// Returns Err if the contents of the storage can not be parsed.
+    pub fn attach_or_create_with_serialization(storage: &'a mut S, _ser: Ser) -> StdResult<Self> {
+        if let Some(len_vec) = storage.get(LEN_KEY) {
+            let off_vec = storage.get(OFFSET_KEY).unwrap();
+            Self::new(storage, &len_vec,&off_vec)
+        } else {
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: DONE, also need to start to implement this way on my contracts too!

File: packages/storage/src/deque_store.rs [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+    
+    /// Add an item to the begining of the collection.
+    ///
+    /// This operation has a constant cost.
+    pub fn push_front(&mut self, item: &T) -> StdResult<()> {
+        self.set_offset(self.off.overflowing_sub(1).0);
+        self.set_at_unchecked(0_u32, item)?;
+        self.set_length(self.len + 1);
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Done

File: packages/storage/src/deque_store.rs [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+                    self.set_at_unchecked(i, &self.get_at_unchecked(i+1).unwrap());
+                }
+            } else { // closer to the head
+                for i in (0..pos).rev() {
+                    self.set_at_unchecked(i+1, &self.get_at_unchecked(i).unwrap());
+                }
+                self.set_offset(self.off.overflowing_add(1).0);
+            }
+            self.set_length(self.len - 1);
+            item
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Added this:
     /// Remove an element off the collection at the position provided
    ///
    /// Removing an element from the head (first) or tail (last) has a constant cost.
    /// Removing from the middle the cost will depend on the proximity to the head or tail.
    /// In this case, all the elements between the closest tip of the collection (head or tail)
    /// and the remove position will be shifting positions.
    ///
    /// Worst case scenario, in terms of cost, will be if the element position is
    /// exactly in the middle of the collection.

File: packages/storage/src/deque_store.rs [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+        }
+    }
+
+    /// Remove element at position provided
+    pub fn remove(&mut self, pos: u32) -> StdResult<T> {
+        if pos == 0 { // if remove the first element
+            self.pop_front()
+        } else if pos == self.len - 1 { // if remove the last element
+            self.pop_back()
+        } else if pos > 0 && pos < self.len - 1 { // if remove element from the middle
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Done

File: packages/storage/src/deque_store.rs [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+            item_type: PhantomData,
+            serialization_type: PhantomData,
+            len,
+            off
+        })
+    }
+
+    pub fn off(&self) -> u32 {
+        self.off
+    }
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Done

File: packages/storage/src/deque_store.rs [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+    type Item = StdResult<T>;
+    type IntoIter = Iter<'a, T, S, Ser>;
+
+    fn into_iter(self) -> Iter<'a, T, S, Ser> {
+        let end = self.len;
+        Iter {
+            storage: self,
+            start: 0_u32,
+            end: end
+        }
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Nice tip. Thanks!

File: packages/storage/src/lib.rs [hunk 8]
```diff
@@ -1,5 +1,7 @@
 pub mod append_store;
+pub mod deque_store;
 pub mod typed_store;
 
 pub use append_store::{AppendStore, AppendStoreMut};
-pub use typed_store::{TypedStore, TypedStoreMut};
+pub use deque_store::{DequeStore, DequeStoreMut};
+pub use typed_store::{TypedStore, TypedStoreMut};
```
[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/lib.rs]: Done

File: packages/storage/src/deque_store.rs (hunk 1 above)
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs:577]: Done

File: packages/storage/src/deque_store.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (reuvenpo) [reply packages/storage/src/deque_store.rs]: Yup, that's almost perfect. If it's ok though, I'll rephrase it a little:
```rust
/// Remove an element from the collection at the specified position.
///
/// Removing an element from the head (first) or tail (last) has a constant cost.
/// The cost of removing from the middle will depend on the proximity to the head or tail.
/// In that case, all the elements between the closest tip of the collection (head or tail)
/// and the specified position will be shifted in storage.
///
/// Removing an element from the middle of the collection 
/// has the worst runtime and gas cost.
```

[the remark this answers is not in the document]
Comment (FSoaresDev) [reply packages/storage/src/deque_store.rs]: Nice, thanks!
Done.

**Review by reuvenpo — APPROVED** [review]
Great work on this PR! Thanks again for submitting it!
It looks great now :)

**Comment by reuvenpo** [comment]
Ah, sorry, one change request!
Please run thee commands:
```sh
git rebase --onto master debug-print deque_store-debug-print
git push --force
```
This will copy your commits and branch to come out of the latest master.
Then, change this pull request to point to `enigmampc:master`.
Once that is done i'll merge this PR :)

**Comment by FSoaresDev** [comment]
Thanks, I've done what you described on the previous comment.

**Pull request merged**